pkg/task/task.go

```diff
@@ -0,0 +1,132 @@
+package task
+
+import (
+	"context"
+	"fmt"
+	"net"
+	"net/url"
+	"syscall"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/rs/xid"
+	v1 "github.com/tinkerbell/pbnj/pkg/api/v1"
+	"github.com/tinkerbell/pbnj/pkg/logging"
+	"github.com/tinkerbell/pbnj/pkg/oob"
+	"github.com/tinkerbell/pbnj/pkg/repository"
+)
+
+// Task interface for doing BMC actions
+type Task interface {
+	Execute(ctx context.Context, logger logging.Logger, description string, action func(chan string) (string, *oob.Error)) (id string, err error)
+	Status(ctx context.Context, logger logging.Logger, id string) (record repository.Record, err error)
+}
+
+// Runner for executing a task
+type Runner struct {
+	Repository repository.Actions
+}
+
+// Execute a task, update repository with status
+func (r *Runner) Execute(ctx context.Context, logger logging.Logger, description string, action func(chan string) (string, *oob.Error)) (id string, err error) {
+	rawID := xid.New()
+	id = rawID.String()
+	l := logger.GetContextLogger(ctx)
+	l.V(0).Info("executing task", "taskID", id, "taskDescription", description)
+	go r.worker(ctx, logger, id, description, action)
+	return id, err
+}
+
+// does the work, updates the repo record
+// TODO handle retrys
+func (r *Runner) worker(ctx context.Context, logger logging.Logger, id string, description string, action func(chan string) (string, *oob.Error)) {
+	l := logger.GetContextLogger(ctx)
+	l.V(0).Info("starting worker", "taskID", id, "description", description)
+	resultChan := make(chan string, 1)
+	errMsgChan := make(chan oob.Error, 1)
+	messagesChan := make(chan string)
+	defer close(resultChan)
+	defer close(messagesChan)
+	repo := r.Repository
+	sessionRecord := repository.Record{
+		StatusResponse: &v1.StatusResponse{
+			Id:          id,
+			Description: description,
+			State:       "running",
+			Messages:    []string{},
+			Error: &v1.Error{
+				Code:    0,
+				Message: "",
+				Details: nil,
+			},
+		}}
+
+	err := repo.Create(ctx, id, sessionRecord)
+	if err != nil {
+		// TODO how to handle unable to create record; ie network error, persistence error, etc?
+		l.V(0).Error(err, "creating record failed")
+		return
+	}
+
+	go func() {
+		for msg := range messagesChan {
+			l.V(0).Info("STATUS MESSAGE", "statusMsg", msg)
+			currStatus, _ := repo.Get(ctx, id)
+			sessionRecord.Messages = append(currStatus.Messages, msg)
+			_ = repo.Update(ctx, id, sessionRecord)
+		}
+	}()
+
+	go func(rChan chan string, sChan chan string, mChan chan oob.Error) {
+		result, errMsg := action(messagesChan)
+		rChan <- result
+		mChan <- *errMsg
+	}(resultChan, messagesChan, errMsgChan)
+
+	sessionRecord.Result = <-resultChan
+	errMsg := <-errMsgChan // nolint
+	// Allow any final status messages to written
+	// TODO better way than this??
+	time.Sleep(2 * time.Second)
```

Comment:
I was losing status messages because although the messages are written to the messagesChan before the resultChan and errMsgChan finish, the goroutine to update the persistence record with the messages hadn't finished. the messages are eventually consistent, but the task response would say complete and not have all the messages. 

Sorry, I hope that makes sense.

Reply:
Ah, I see, so the problem wasn't with these lines. I didn't realize that when I was reading through. That makes sense. I think perhaps a better solution would be to have a status done channel that the go func could write to once the channel it's reading on is closed, if that makes sense. Perhaps something like

```
status_done := make(chan bool, 1)
go func(){
  for msg := range other_chan { fmt.Println(msg); }
  status_done <- true
}()
<-status_done
fmt.Println("completed")
```

or something along those lines... just a potential alternative solution. I think this is also assuming that `other_chan` would get closed which would cause the range to quit. If that doesn't happen then this might not be the best solution.

However with that said, you could probably close it after you read errMsg and then wait for status_done...

Anywho, just some ideas though :)

Reply:
good stuff, thanks for the ideas. I used them and update the PR. Let me know what you think, thanks.